### PR TITLE
refactor(cli): adding prompts for pk key and type for relation cli

### PR DESCRIPTION
**Pull request by agnes512**
connects to #6017 and #6131 

In spike 6017 we realized that current relation CLI frequently reads files, which causes lots of overhead. Apart from that, as it is reported in 6131, the current CLI couldn't handle the inheritance well because it checks the primary keys by **reading files** instead of models themselves. 

Therefore, this PR changes the way the relation CLI gets the name and the type of pks by adding more prompts, which is similar to what we do for the controller (suggested by @Salketer). Even though the change adds a bit complexity, it will speed up the relation tests, and the inheritance issue won't exist anymore

#### Before
```
$ lb4 relation
? Please select the relation type hasMany
? Please select source model TodoList
? Please select target model Todo
? What is the name of ID property of the source model? id
? Source property name for the relation getter (will be the relation name) todos
? Allow TodoList queries to include data from related Todo instances? Yes
```

#### After
```
$ lb4 relation
? Please select the relation type hasMany
? Please select source model TodoList
? Please select target model Todo
? What is the name of ID property of the source model? id
? What is the type of the source model primary key? number
? What is the name of ID property of the target model? id
? What is the type of the target model primary key? number
? Foreign key name to define on the target model todoListId
? Source property name for the relation getter (will be the relation name) todos
? Allow TodoList queries to include data from related Todo instances? Yes
```

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: docs/site/Relation-generator.md [hunk 1]
```diff
@@ -61,6 +73,17 @@ lb4 relation --sourceModel=<sourceModel>
 
 - `<destinationModel>` - Name of the model to create a relationship with.
 
+- `<sourceModelPrimaryKey>` - Name of the id property on the model to create the
+  relationship from.
+
+- `<sourceModelPrimaryKeyType>` - Type of the id property on the source mode.
```
Comment (nabdelgadir) [inline docs/site/Relation-generator.md:79]:
```suggestion
- `<sourceModelPrimaryKeyType>` - Type of the id property on the source model.
```

File: docs/site/Relation-generator.md [hunk 2]
```diff
@@ -61,6 +73,17 @@ lb4 relation --sourceModel=<sourceModel>
...
+
+- `<destinationModelPrimaryKey>` - Name of the id property on the model to
+  create a relationship with.
+
+- `<destinationModelPrimaryKeyType>` - Type of the id property on the
+  destination mode.
```
Comment (nabdelgadir) [inline docs/site/Relation-generator.md:85]:
```suggestion
  destination model.
```

File: docs/site/Relation-generator.md [hunk 3]
```diff
@@ -121,6 +149,24 @@ The tool will prompt you for:
   to choose from as the through model of the relation. For HasManyThrough
   relation only.
 
+- Name of the id property of the source model (`sourceModelPrimaryKey`). Prompts
+  for the name of the source model. For example, for `Customer` has many
+  `Order`s, `Customer` is the source model. For an `Order` belongs to a
+  `Customer`, `Order` is the source model. Defaults value: `id`.
+
+- Type of the id property of the source model (`sourceModelPrimaryKey`). Prompts
```
Comment (nabdelgadir) [inline docs/site/Relation-generator.md:157]:
```suggestion
- Type of the id property of the source model (`sourceModelPrimaryKeyType`). Prompts
```

File: docs/site/Relation-generator.md [hunk 4]
```diff
@@ -121,6 +149,24 @@ The tool will prompt you for:
...
+  a list of types fir the id property on the source model. Default value:
```
Comment (nabdelgadir) [inline docs/site/Relation-generator.md:158]:
```suggestion
  a list of types for the id property on the source model. Default value:
```

File: docs/site/Relation-generator.md [hunk 5]
```diff
@@ -121,6 +149,24 @@ The tool will prompt you for:
...
+  `number`.
+
+- Name of the id property of the source model (`destinationModelPrimaryKey`).
```
Comment (nabdelgadir) [inline docs/site/Relation-generator.md:161]:
```suggestion
- Name of the id property of the destination model (`destinationModelPrimaryKey`).
```

File: docs/site/Relation-generator.md [hunk 6]
```diff
@@ -121,6 +149,24 @@ The tool will prompt you for:
...
+  Prompts for the name of the destination model. For example, for `Customer` has
+  many `Order`s, `Order` is the destination model. For an `Order` belongs to a
+  `Customer`, `Customer` is the destination model. Defaults value: `id`.
```
Comment (nabdelgadir) [inline docs/site/Relation-generator.md:164]:
```suggestion
  `Customer`, `Customer` is the destination model. Default value: `id`.
```

File: docs/site/Relation-generator.md [hunk 7]
```diff
@@ -121,6 +149,24 @@ The tool will prompt you for:
...
+
+- Type of the id property of the source model (`sourceModelPrimaryKey`). Prompts
+  a list of types fir the id property on the source model. Default value:
+  `number`.
+
```
Comment (nabdelgadir) [inline docs/site/Relation-generator.md:169]:
```suggestion
- Type of the id property of the destination model (`destinationModelPrimaryKeyType`). Prompts
  a list of types for the id property on the source model. Default value:
  `number`.
```

File: docs/site/Relation-generator.md [hunk 8]
```diff
@@ -86,7 +109,12 @@ HasMany, HasOne, or BelongsTo relations.
 lb4 relation --sourceModel=<sourceModel>
 --destinationModel=<destinationModel> --throughModel=<throughModel>
 --relationType=<hasManyThrough> [--relationName=<relationName>]
+[--sourceModelPrimaryKey=<sourceModelPrimaryKey>]
```
Comment (nabdelgadir) [inline docs/site/Relation-generator.md:112]: For `HasManyThrough`, is `sourceModelPrimaryKey` the same as `sourceKeyOnThrough` and `destinationModelPrimaryKey` the same as `targetKeyOnThrough`?

**Review by nabdelgadir — COMMENTED** [review]
I am a bit concerned about the user experience just because this a lot more prompts to go through, but I think the benefits you mentioned could be a good tradeoff? Either way, good job! I just have some minor nitpicks. 😄

File: docs/site/Relation-generator.md (hunk 8 above)
Comment (agnes512) [reply docs/site/Relation-generator.md:112]: Good question! Indeed they refer to the same thing, but they may be different. For example, `sourceModelPrimaryKey` on `Customer` is `id`, and `sourceKeyOnThrough` on `CustomerOrderLink` defaults to `customerId`, and `targetKeyOnThrough` defaults to `orderId`. Therefore those two params are needed IMO.

**Comment by dhmlau** [comment]
_Just commenting on the before/after user experience_
From the user experience perspective, it seems to be redundant to ask about the property type where the previous prompt asks to select a property.  It might be along the same line as https://github.com/strongloop/loopback-next/issues/1789, where we ask about the id then the type in the controller cli.

**Comment by Salketer** [comment]
The worse usage scenario is having the CLI fail saying there is no id detected... Starting from there, we could have the prompts only when they cannot be guessed. This would not add any friction, just reduce some.

As from what I understand, you'd want to reduce file reads and, to do so, always prompt the user... That of course will add friction. Is it really important to reduce the file reads? I may not have the full picture here, but the only time I recall reading complains about this is because tests take a lot of time because of it... It's still faster to run the test than to answer a prompt.

**Comment by agnes512** [comment]
Thanks for all the comments. I open a new PR https://github.com/strongloop/loopback-next/pull/6937 based on suggestions. Please take a look! cc @Salketer

**Pull request closed without merging**